test/unit/math/test_ad.hpp

```diff
@@ -1432,6 +1433,54 @@ void expect_ad_vectorized_binary(const F& f, const T1& x, const T2& y) {
   expect_ad_vectorized_binary(tols, f, x, y);
 }
 
+template <typename T1, typename T2,
+          require_all_stan_scalar_t<T1, T2>* = nullptr>
+void expect_float_equal(const T1& x, const T2& y) {
+  EXPECT_FLOAT_EQ(x.val(), y.val());
+  EXPECT_FLOAT_EQ(x.adj(), y.adj());
+}
+template <
+    typename T1, typename T2,
+    require_any_eigen_t<T1, value_type_t<T1>, T2, value_type_t<T2>>* = nullptr>
+void expect_float_equal(const T1& x, const T2& y) {
+  EXPECT_MATRIX_FLOAT_EQ(x.val(), y.val());
+  EXPECT_MATRIX_FLOAT_EQ(x.adj(), y.adj());
+}
+
+/**
+ * For an unary function check that an Eigen matrix of vars and a var with an
+ * inner eigen type return the same values and adjoints. This is done by
+ * calling the function on both types, summing the results, then taking the
+ * gradient of the sum which will propogate the adjoints upwards.
+ *
+ * @tparam F A lambda or functor type that calls the unary function.
+ * @tparam T An Eigen matrix type
+ * @param f a lambda
+ * @param x An Eigen matrix.
+ *
+ */
+template <typename F, typename T>
+void expect_ad_matvar(F&& f, const T& x) {
+  using stan::plain_type_t;
```

Comment:
This checks that `var<matrix>` and `matrix<var>` return the same values and give the same gradients.

I think it should also check that it has the same throws as the other stuff.

In the future we'll need multiple argument testing as well (which means some args are autodiff and some are primitive). Are you going to do that in another pull? We should probably test mixes of `var<mat>` `mat<var>` in that case too.

Right now I think it makes sense that the entry-point to this testing is separate from the rest of the testing framework, but the Spirit of the Testing Framework is that all tests run automatically, so I think this should become part of the regular `expect_ad` function eventually.

Reply:
> This checks that var<matrix> and matrix<var> return the same values and give the same gradients.

> I think it should also check that it has the same throws as the other stuff.

Agree adding that now!

> In the future we'll need multiple argument testing as well (which means some args are autodiff and some are primitive). Are you going to do that in another pull? We should probably test mixes of var<mat> mat<var> in that case too.

Yeah I'm building these as we need them. I don't want to try to future proof now without examples

> Right now I think it makes sense that the entry-point to this testing is separate from the rest of the testing framework, but the Spirit of the Testing Framework is that all tests run automatically, so I think this should become part of the regular expect_ad function eventually.

Once we have a binary op in I'm going to start looking at how to do this. I wasn't sure how to automatically do this at first but I have an idea now. We can just detect whether the signature `f(var<mat>)` exists and only run these tests in that instance

Reply:
> We can detect whether the signature f(var<mat>) exists and only run these tests in that instance

Ack, actually I'm not sure how to do this. The issue is that we to have a function that only calls the tests when the function has an overload that accepts a `var<matrix>`

```cpp
// if F(var<matrix>) is a valid expression do the tests
template <typename F, typename T, require_t<is_overloaded<F, var<matrix>>>* = nullptr>
void expect_ad_matvar(F&& f, const T& y) {
  // do the tests
}

// if F(var<matrix>) is a not valid expression don't do the tests
template <typename F, typename T, require_t<is_not_overloaded<F, var<matrix>>>* = nullptr>
void expect_ad_matvar(F&& f, const T& y) {}
```

But in our test suite `F` is always a lambda with `auto` as the argument type like

```cpp
auto test_lambda_F = [](const auto& x) { return actual_func(x);};
```

 so `F(var<matrix>)` is a valid expression and passes SFINAE here because of the `auto`. But when the compiler tries to do resolution it finds that it can't compile the body of the lambda and errors.

There's an example here where I'm just doing this with `double` (equivalent of `Matrix<var>`) and tries to go to the failure overload for an input of `std::string`

https://godbolt.org/z/o7EW8E

Reply:
> so F(var<matrix>) is a valid expression and passes SFINAE here because of the auto.

Ugh. Not obvious to me how to get around this. I guess it'll just be our responsibility to turn on `var_value` testing for functions where it is implemented.

Since we aren't doing explicit casts at some point we'll have to be tracking what functions support `var<mat>` types and what functions can only do `mat<var>`, so there is already a problem at least as complicated as this one in existence.

Reply:
Yeah let's re-approach this once we get closer to full coverage

Reply:
You can do this by having the third parameter of the function be ill-formed if the input/s don't meet particular constraints. 
I've modified your code to show this in action: https://godbolt.org/z/ejP9Mn

Where the ```whatsup``` functions are given a third parameter that applies a function that would only be appropriate for the specified type, otherwise its ill-formed.

From templates to parameters, it's SFINAE all the way down!

Reply:
I think the requirement is that the function `F` take in a certain type, not that the `y` is of a certain type. This is different from the thing you posted right?

Reply:
I was trying something similar to that before, the issue is that the lambda always passes SFINAE since it's argument is `auto`. The compiler fails after SFINAE while processing the body of the function. 

https://godbolt.org/z/nj3rn7

The only scheme I've been able to work out is to actually pass the function as well so that type deduction can be done directly on the function we want to test `var<mat>` vs. `mat<var>` for

https://godbolt.org/z/brGofz

Reply:
Oh got it, got it. After some digging looks like an easy fix is to have a trailing return type for the lambda:
```
int main() {
  whatsup([](auto x) -> decltype(foo(x)) { foo(x);}, 2.0);
  whatsup([](auto x) -> decltype(foo(x)) { foo(x);}, std::string("UUggghhh!"));
  return 0;
}
```

Working here: https://godbolt.org/z/TeaMde

Reply:
This seems like a good solution to me given the tests won't compile without this. I would prefer to do the unified `mat<var>` `var<mat>` test work in another pull though to not stall the other stuff in this branch.